typing/typecore.ml

```diff
@@ -4770,247 +4773,334 @@ let spellcheck_idents ppf unbound valid_idents =
+    else if Path.same expected_type Predef.path_float then
       hint_suffix '.'
     else
-      None
-  in
-  let hint str_val =
-    match h with
-    | Ctype.Unification_trace.(Diff
-      { expected = { t = { desc = Tconstr (typ, [], _); _ }; _ }; _ }) ->
-        hint_const typ str_val
-    | _ -> None
+      ()
   in
   match const with
   | Const_int n -> hint (Int.to_string n)
   | Const_int32 n -> hint (Int32.to_string n)
   | Const_int64 n -> hint (Int64.to_string n)
   | Const_nativeint n -> hint (Nativeint.to_string n)
-  | _ -> None
+  | _ -> ()
 
-let report_literal_type_constraint ppf const trace =
-  let hints ~prev:_ h = report_literal_type_constraint h const in
-  begin match Ctype.Unification_trace.explain trace hints with
-  | Some hint -> hint ppf
-  | None -> ()
-  end
+let report_literal_type_constraint ppf const = function
+  | Some Unification_trace.
+    { expected = { t = { desc = Tconstr (typ, [], _) } } } ->
+      report_literal_type_constraint ppf typ const
+  | Some _ | None -> ()
 
-let report_expr_type_clash_hints ppf exp trace =
+let report_expr_type_clash_hints ppf exp diff =
   match exp with
-  | Some (Texp_constant const) -> report_literal_type_constraint ppf const trace
+  | Some (Texp_constant const) -> report_literal_type_constraint ppf const diff
   | _ -> ()
 
-let report_pattern_type_clash_hints ppf pat trace =
+let report_pattern_type_clash_hints ppf pat diff =
   match pat with
-  | Some (Tpat_constant const) -> report_literal_type_constraint ppf const trace
+  | Some (Tpat_constant const) -> report_literal_type_constraint ppf const diff
   | _ -> ()
 
+(* Hint when using int's operators (eg. `+`)
+   on other kind of integer and floats *)
+let report_numeric_operator_clash_hints actual_type operator =
+  let stdlib = Path.Pident (Ident.create_persistent "Stdlib") in
+  let stdlib_p mod_ val_ = Path.Pdot (Path.Pdot (stdlib, mod_), val_) in
+  let hint expected_op =
+    Some (fun ppf ->
+      fprintf ppf "@[Hint:@ Did you mean to use `%a'?@]"
+        Printtyp.path expected_op
+    )
+  in
+  let hint ~add ~sub ~mul ~div ~mod_ () =
+    let is_op op = Path.same operator (Path.Pdot (stdlib, op)) in
+    if is_op "+" then hint add
+    else if is_op "-" then hint sub
+    else if is_op "*" then hint mul
+    else if is_op "/" then hint div
+    else if is_op "mod" then hint mod_
+    else None
+  in
+  let hint_qualified name =
+    let p = stdlib_p name in
+    hint ~add:(p "add") ~sub:(p "sub") ~mul:(p "mul")
+      ~div:(p "div") ~mod_:(p "rem") ()
+  in
+  let expecting = Path.same actual_type in
+  if expecting Predef.path_int32 then
+    hint_qualified "Int32"
+  else if expecting Predef.path_int64 then
+    hint_qualified "Int64"
+  else if expecting Predef.path_nativeint then
+    hint_qualified "Nativeint"
+  else if expecting Predef.path_float then
+    let p id = Path.Pdot (stdlib, id) in
+    hint ~add:(p "+.") ~sub:(p "-.") ~mul:(p "*.") ~div:(p "/.")
```

Comment:
This hint should be `hint_std ()` defined before `let hint_qualified name` -- "do not mix abstraction levels", this if..else if..else nest is mixing function calls with raw expressions.

Reply:
Yes, this is a great increase in readability.